subprojects/core/src/main/java/org/gradle/initialization/buildsrc/BuildSourceBuilder.java

```diff
@@ -115,45 +112,29 @@ public String getBuildPath() {
         });
     }
 
-    private ClassPath buildBuildSrc(BuildDefinition buildDefinition) {
-        final BuildController buildController = createBuildController(buildDefinition);
-        try {
-            buildController.getGradle().rootProject(new Action<Project>() {
-                @Override
-                public void execute(Project rootProject) {
-                    buildRegistry.addNestedBuild(buildController.getGradle().getSettings());
+    private ClassPath buildBuildSrc(final BuildDefinition buildDefinition) {
+        NestedBuildState nestedBuild = buildRegistry.addNestedBuild(buildDefinition, nestedBuildFactory);
+        return nestedBuild.run(new Transformer<ClassPath, BuildController>() {
+            @Override
+            public ClassPath transform(BuildController buildController) {
+                GradleInternal build = buildController.getGradle();
+                StartParameter startParameter = buildController.getGradle().getStartParameter();
+                if (build.getParent().findIdentityPath() == null) {
+                    // When nested inside a nested build, we need to synthesize a path for this build, as the root project is not yet known for the parent build
```

Comment:
It would be great to extract methods for things like this. Inline comments break the flow and quickly go stale during refactorings.

```
/**
  * Explanation here
  */
private void ensureBuildHasPath() {...}
```